### ISSUE TITLE
"global_valid_tokens" scales badly

### ISSUE THREAD

**Issue by rakkit** [pytorch/torchtitan#2850]
### Bug description

for now we pre-fetch all micro batchsize and run all_reduce to get "global_valid_tokens". 
i found these two ops
`local_valid_tokens += (labels != IGNORE_INDEX).sum()`
  ` dist_utils.dist_sum(local_valid_tokens, batch_mesh)`
have bad performance at scale. @wwwjn 

The setup is  ~8B moe,  local batch size 9 x gradient accumulation 3,  runs on HSDP where FSDP fixed to 64

<img width="1316" height="884" alt="Image" src="https://github.com/user-attachments/assets/1c4e11d8-2943-49aa-a032-9d6f59bf09c9" />

For reference, before this refactor (in December) its runs on 12k~13k tps smoothly on 384 gpus.

There are two things slow down the training, one is micrio batch prefetching, this takes up to ~350ms (end2end   steps takes 8000ms so this data load waste 4.5% time)

<img width="1178" height="270" alt="Image" src="https://github.com/user-attachments/assets/cb2009d8-9d90-4fc9-8bd9-094057a66dbc" />

And second (main bottleneck) is this `dist_sum` for global tokens count.  this scaling badly with more gpus.  e.g. on 128 gpus, this takes ~302.981 ms and on 256 gpus, its takes ~1,279.202 ms
(its launched `ncclDevKernel_AllReduce_Sum_u64_TREE_LL(ncclDevKernelArgsStorage<4096ul>)` kernles FYI) 

(correction, i somehow set `EOS` in labels to be `IGNORE_INDEX` in my local branch, so my training setup actaully have lots of `IGNORE_INDEX`, i think thats why this `ne` and `sum` takes long to compute) 

### Versions

PyTorch nightly (`torch==2.12.0.dev20260401`) + CU129

**Comment by rakkit** [comment]
ad-hoc solution is for llm-pretrain side, we could do fast-path `global_valid_tokens=local_valid_tokens*dp-size` to skip both of them.  i will think about if there is any other solutions

**Comment by tianyu-l** [comment]
> one is micrio batch prefetching, this takes up to ~350ms (end2end steps takes 8000ms so this data load waste 4.5% time)

Where is the time mostly spent? torchtitan dataloader is pretty toy-ish so it's not a surprise that some bottleneck would exist, but I would love to know more details. We may consider swapping the dataloader to something else. cc @joecummings 

> And second (main bottleneck) is this `dist_sum` for global tokens count. this scaling badly with more gpus. e.g. on 128 gpus, this takes ~302.981 ms and on 256 gpus, its takes ~1,279.202 ms

Oh it sounds that NCCL all-reduce takes 4x time when we 2x the world size. cc @fegin

**Comment by rakkit** [comment]
> Where is the time mostly spent?

`local_valid_tokens += (labels != IGNORE_INDEX).sum().float()`   
there gonna be `gradient_accumulation_steps` x (`ne` and `sum` )

in my setup (9 local bs x 4096 seqlen,  3 accumulation), this `local_valid_tokens` calculate takes 350ms

(the pre-fetching i mean what we did in trainer, not on actual dataloader side) 


> it sounds that NCCL all-reduce takes 4x time when we 2x the world size

 yes, i think this might because it use tree reduce. 


i am testing the fast-path now, (for llm-pretraining we can just skip this calculation and reduce) this is the easiest i can do on my side

**Comment by tianyu-l** [comment]
https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L742 is happening on CPU, so we should be able to do async prefetching on this, so that GPU is not blocked. WDYT?

The fast-path doesn't work for any training with ignored (-100) labels, whether it's SFT or vision-language models.

**Comment by rakkit** [comment]
> The fast-path doesn't work for any training with ignored (-100) labels, whether it's SFT or vision-language models.

yes i agree, just i need smth immediacy can fix it for LLM pre-train.


> Oh it sounds that NCCL all-reduce takes 4x time when we 2x the world size. cc @fegin

now its becomes more weird. 

the tps did not get imporvement even we skip the data load 

<img width="1316" height="884" alt="Image" src="https://github.com/user-attachments/assets/93e0954a-60d5-43dc-8053-84c12cd2511a" />

from profiler, now the all-gather of embedding get slower and slower when world size increases.  (before it was actually ~10ms) and now its 300ms on 64 gpus and 5,676.849 ms on 384 gpus.     i think there are smth really werid there blocked first communication.  i will run test on differnt machines see if its a system error

**Comment by rakkit** [comment]
right, i found multiple things that affect the measurement. 
1) in my local branch, i set all `EOS` in labels to be `IGNORE_INDEX`, and this effectively somehow make my previous test more like "sft".
2) "NCCL all-reduce takes 4x time when we 2x the world size." and "all-gather of embedding get slower and slower when world size increases." is some kind of hallucination maybe machine unstable or maybe its about profiler.



I re-run the test and revert my local changes (EOS is just EOS,  and in princple no `IGNORE_INDEX` in lable) and turn off profile. 

first machine (HBM2 h100)

<img width="1385" height="845" alt="Image" src="https://github.com/user-attachments/assets/4a0eb82d-cb9a-4a32-86e5-bbb4865a1ba7" />


another machine (normal h100)





<img width="1385" height="846" alt="Image" src="https://github.com/user-attachments/assets/2f09a29c-ece3-4d1d-8efb-02f397132205" />


And the results looks pretty much make sense now.      current default behaviors might get slightly slow on LLM-pretraining, (when there is no `IGNORE_INDEX`).


And  in practice, it matters a lot for SFT and VLM training, to be specific, anything that have lots of "IGNORE_INDEX" in labels

**Comment by wwwjn** [comment]
Thanks for reporting and experiment around it, if the results turns out make sense now, temporarily close this issue

P.S. #2937  also refactor the loss scaling part slightly, which should not expected to change your experiment results. #

**Comment by rakkit** [comment]
@wwwjn em..  it only works OK for pre-training parts (case we don't have IGNORE_INDEX in label), once move to SFT (or any task that have lots of IGNORE_INDEX) the bottleneck is still there

**Comment by wwwjn** [comment]
I see, the AllReduce will introduce some slowdown in trade of numeric correctness. One idea to relief this slowdown was @fegin  was planning to overlap the data loading process at N step with previous N-1  step's fwd/bwd, basically prefetching the data.
